Connect suggestions frontend and backend

AllanYangZhou opened the pull request:
Addresses #1291 and #1304, minus the e2e tests.

seanlip commented on core/domain/exp_services.py:
```diff
@@ -1500,7 +1500,7 @@ def accept_suggestion(editor_id, thread_id, exploration_id, commit_message):
              full_commit_message)
         feedback_services.create_message(
-            exploration_id, thread_id, suggestion_author_userid, None, None,
+            exploration_id, thread_id, editor_id, None, None,
```
Ooh, good catch!

Also, it's an unrelated nit, but could you change suggestion_author_userid to suggestion_author_id elsewhere in this method, just to be consistent (since we don't use 'userid' anywhere)? The alternative is suggestion_author_user_id but that's a bit long.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -96,19 +96,25 @@ oppia.controller('FeedbackTab', [
+        var states = explorationData.data.states;
+        $scope.oldContent = states[suggestion.state_name].content[0].value;
+        $scope.newContent = suggestion.state_content.value;
+        $scope.commitMessage = {
```
The 'value' thing is legacy, I believe. I don't think we need to replicate it here, a string should do. (Unless you were thinking of adding more data to $scope.commitMessage in the future?)

AllanYangZhou replied:
Ah, this was actually a workaround because for some reason I put the corresponding view code (the input element for the commit message) in an "ng-if." I've changed the ng-if to ng-show and am just using a string now.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -121,20 +127,31 @@ oppia.controller('FeedbackTab', [
           $modalInstance.dismiss('cancel');
         };
+
+        $scope.suggestionValid = function() {
```
isSuggestionValid

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -121,20 +127,31 @@ oppia.controller('FeedbackTab', [
+            $scope.setActiveThread($scope.activeThread.thread_id);
+          });
+          // Immediately update editor to reflect accepted suggestion.
+          if(result.action === 'accept') {
```
nit: add single space after 'if'

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -96,19 +96,25 @@ oppia.controller('FeedbackTab', [
         $scope.acceptSuggestion = function() {
           $modalInstance.close({
-            action: 'accept'
+            action: 'accept',
```
Pull 'accept' out into a named constant: ACTION_ACCEPT_SUGGESTION, perhaps (since it's also used in line 144). Ditto 'reject', for consistency.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -121,20 +127,31 @@ oppia.controller('FeedbackTab', [
+            var state = angular.copy(explorationData.data.states[stateName]);
+            state.content[0].value = suggestion.state_content.value;
+            explorationStatesService.setState(stateName, state);
+            $rootScope.$broadcast('refreshStateEditor');
```
I just realized an issue with this. What if the creator is editing an exploration, and then accepts a suggestion (before saving their draft)? After accepting the suggestion, they won't be able to save their existing changes because the version number will be wrong. (Could you confirm this is an issue, too?)

It's not great, but I think a workaround for now would be to prevent the creator from accepting suggestions until they've saved any drafts they currently have. Maybe we can display a message along the lines of "Please save your draft changes before accepting this suggestion." It's not ideal at all, unfortunately, but I think it will have to do for now.

AllanYangZhou replied:
I've implemented a check to prevent acting on suggestions when there are unsaved changes: see be0fd91

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -121,20 +127,31 @@ oppia.controller('FeedbackTab', [
+            $rootScope.$broadcast('refreshStateEditor');
+          }
+        }, function(res) {
+          console.log("Error resolving suggestion");
```
nit: we use single quotes in JS

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/ThreadDataService.js:
```diff
@@ -31,18 +31,24 @@ oppia.factory('threadDataService', [
+      // Later requests will use only the thread_id, not full_thread_id
+      _data.feedbackThreads.concat(_data.suggestionThreads).map(function(val) {
+        val.thread_id = val.full_thread_id.substring(
+          val.full_thread_id.indexOf(".") + 1, val.full_thread_id.length);
```
nit: single quotes

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/ThreadDataService.js:
```diff
@@ -117,10 +125,15 @@ oppia.factory('threadDataService', [
+      if (commitMsg) {
+        payload.commit_message = commitMsg;
+      }
+      $http.put(_SUGGESTION_ACTION_HANDLER_URL+threadId, payload).success(
```
nit: single spaces around '+'

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -279,6 +279,12 @@ oppia.factory('oppiaPlayerService', [
+      return version;
+    },
+    getStateContent: function(stateName) {
+      return exploration.getState(stateName).content;
```
Can this be refactored slightly to:

    getStateContentHtml: function(stateName) {
      return exploration.getContentHtml(stateName);
    }

where the result is a string? I think all the [0] and 'value' and '.content' stuff should be encapsulated within the Exploration object.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -40,15 +40,50 @@
+
+    <div class="modal-body">
+      <text-angular-rte ng-show="showEditor" html-content="suggestionContent"></text-angular-rte>
+      <br />
```
Nit: This tag is self-closing, so just `<br>` will do.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -40,15 +40,50 @@
+
+    <div class="modal-footer">
+      <button class="btn btn-default" ng-click="cancelSuggestion()">Cancel</button>
+      <button class="btn btn-success" ng-click="submitSuggestion()" ng-disabled="initContent==suggestionContent">Submit</button>
```
Nit: put single spaces around '=='.

seanlip replied:
Also, maybe call it "Submit Suggestion"

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -40,15 +40,50 @@
-        </li>
-      {% endif %}
+        {% elif username %}
+          <a ng-click="showLearnerSuggestionModal()" tooltip="Suggest Edit" tooltip-placement="bottom" target="_blank">
```
I suggest "Suggest a Change" as the tooltip. The current tooltip text reads as two separate actions (especially since the content ends up being on two separate lines.)

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -40,15 +40,50 @@
+  </script>
+  <script type="text/ng-template" id="modals/learnerViewSuggestion">
+    <div class="modal-header">
+      <h3>Make a Suggestion</h3>
```
Suggest an Edit (or Suggest a Change)

AllanYangZhou replied:
Done.

seanlip commented:
Thanks, @AllanYangZhou -- this is looking great, and it's very nice to see this in action! Also /cc @sbhowmik89 for info.

A few UI comments, in addition to the code comments above:

- In the editor feedback tab, let's remove the horizontal line between the sections, and separate them with additional space. It looks cleaner that way.
- In the thread view, can the "View Suggestion" button have top and bottom margins of, say, 6 px? It looks a bit crowded at the moment.
- If the suggestion is invalid, rejected or already accepted, can the "View Suggestion" button at the top of the thread be a 'normal' button (similar to the 'Cancel' button in the modal), not blue? Otherwise it stands out too much, I think.
- In the suggestion review modal, can there be a bit more vertical spacing between the error message and the buttons at the bottom?
- When I accept/reject a suggestion, I should see a "Status changed to 'Fixed'" (or 'Ignored') message in the thread view, similar to if I changed the status manually.
- It is possible for me to go through the following flow: reject a suggestion, change status manually to 'Open', and then reject it again. This probably shouldn't be allowed, and the check should be for whether the suggestion itself has been rejected/accepted already, not the current status of the thread.
- I created an exploration, as user A, and then logged in as user B and made two separate suggestions for the same state. I logged in again as user A and went to the editor, accepted one of the suggestions, and rejected the other. However, if I return to either thread and click to view the suggestion, the value shown in the "Current" field does not match the value I had accepted. This looks like a bug...

AllanYangZhou commented:
Hi Sean, thanks for the feedback!
Regarding your last bullet point, I was unable to reproduce the error. Here's what I did for reference:

1. As a learner, I submit 2 suggestions to the same state
1. As an editor, I accept the first suggestion. If I click on "View Suggestion" at this point, I see the changes from the first suggestion in effect in the "Current" panel.
1. I then open the second suggestion, and reject it. If I click "View Suggestion" again, I still see the changes from the first suggestion in the "Current" panel.
1. If I click into the editor view, I see the changes from the first suggestion in effect.

Am I missing any steps?

Regardless, the fact that I am able to reject the second suggestion is a bug, because I was not updating the exploration version after accepting the first suggestion. The correct behavior is, after accepting the first suggestion, the second suggestion becomes invalid because it is now out of date. I've pushed a commit fixing this.

AllanYangZhou commented:
Also, for your 5th bullet point, the message saying "Suggestion {Accepted, Rejected}" is added from the server side. I assume the "Status Changed to..." message would be added from the server side as well, but I'm not really sure how.

For your 6th bullet point, looking at the data received from a suggestion thread, I can't see a way to determine if a suggestion has been acted upon already, besides the status. I could scan through the thread messages looking for a 'Suggestion Accepted' message, but that doesn't seem like a good idea. @sbhowmik89, could a property be added to the thread object that tells whether a suggestion has been acted upon before?

seanlip commented:
Ah, what you did (with respect to my last bullet point) sounds right. Perhaps it got fixed, or I made a mistake -- I'll try playing around with it again!

AllanYangZhou commented:
Yeah, seems I accidentally fixed the bug you described in c0a5776 with the removal of the "angular.copy" (which I removed because I couldn't remember why I put it there). The bug is actually a little more general--the "current" view panel doesn't update after accepting a suggestion if you don't reload the page. For some reason removing the angular.copy fixes this.

AllanYangZhou commented:
@seanlip 
Addressed all bullet points, PTAL.

seanlip commented on core/domain/exp_services.py:
```diff
@@ -1517,7 +1517,6 @@ def reject_suggestion(editor_id, thread_id, exploration_id):
-            'Suggestion rejected.') 
-        thread.status = feedback_models.STATUS_CHOICES_IGNORED
+            exploration_id, thread_id, editor_id,
+            feedback_models.STATUS_CHOICES_FIXED, None, 'Suggestion rejected.') 
```
Should this be STATUS_CHOICES_IGNORED?

AllanYangZhou replied:
Oops! Fixed.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -96,45 +114,76 @@ oppia.controller('FeedbackTab', [
       resolve: {
-        suggestion: function() {
-          return $scope.activeThread.suggestion;
+        suggestionInfo: function() {
```
Suggest breaking this into two separate things -- it's not so good to pass big dict blobs if possible.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -96,45 +114,76 @@ oppia.controller('FeedbackTab', [
+            hasUnsavedChanges: $scope.hasUnsavedChanges()
+          };
+        },
+        stateContent: function() {
```
(In contrast to the previous comment, the key-value pairs in this one are more conceptually related, so I think it's OK as-is, although separate entries would be fine as well.)

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -96,45 +114,76 @@ oppia.controller('FeedbackTab', [
+        $scope.canActOnSuggestion = (
+          suggestionInfo.isSuggestionValid && !suggestionInfo.hasUnsavedChanges
+        );
+        $scope.errorMessage = suggestionInfo.isSuggestionValid
```
Probably more conceptually correct for this to be:

    $scope.errorMessage = (
      !suggestionInfo.isSuggestionValid ? ACTION_INVALID_MSG : 
      suggestionInfo.hasUnsavedChanges ? UNSAVED_CHANGES_MSG : 
      '');

Basically, the JS should make sense without having to know the structure of the DOM.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -96,45 +114,76 @@ oppia.controller('FeedbackTab', [
+        );
+        $scope.errorMessage = suggestionInfo.isSuggestionValid
+          ? UNSAVED_CHANGES_MSG : ACTION_INVALID_MSG;
+        $scope.oldContent = stateContent.oldContent;
```
OK, if you're going to separate them here anyway, you might as well keep them separate in the resolve() section...

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -96,45 +114,76 @@ oppia.controller('FeedbackTab', [
-      }
-      $scope.addNewMessage($scope.activeThread.thread_id, msg, status);
+        $scope.activeThread.thread_id, result.action, result.commitMsg,
+        function(res) {
```
res is unused, so I don't think you need it here.

Also, did you run the JS linter? It should catch this...

AllanYangZhou replied:
Ah, I wasn't aware about the linter. I've fixed this (and other linter issues raised in this file).

seanlip commented on core/templates/dev/head/editor/feedback_tab/ThreadDataService.js:
```diff
@@ -31,18 +31,24 @@ oppia.factory('threadDataService', [
-      _data.threadList = [].concat(res[0].data.threads, res[1].data.threads);
+      _data.feedbackThreads = res[0].data.threads;
+      _data.suggestionThreads = res[1].data.threads;
+      // Later requests will use only the thread_id, not full_thread_id
```
Would it be possible for the backend to supply the frontend with just the thread_id, and not the full_thread_id? As it is, the current setup is a bit awkward.

AllanYangZhou replied:
Yes, I think it would just be moving this processing step to the server, before it returns the threads. Is that what you mean?

seanlip replied:
Yep! If we can confine the use of full_thread_id to as deep in the stack as possible, and not have it percolate upwards, that would be best.

AllanYangZhou replied:
Great, done.

AllanYangZhou commented:
@seanlip I've addressed all the comments. In addition to the comments, I made the following changes:

1. Exploration editors can reject outdated suggestions, if they have not been acted upon before. There are also now separate error messages for suggestions that were previously acted upon, and suggestions that are outdated. See f386c7b.
2. I've disabled adding and editing images from the learner view suggestions. They will just get a message saying they can't edit images if they click the button.  See 9bbadaa.

PTAL.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,237 @@
+    var _isSuggestionCurrent = function() {
+      var activeThread = $scope.activeThread;
+      var suggestionExpVersion = activeThread.suggestion.exploration_version;
+      return explorationData.data.version === suggestionExpVersion;
```
I think this has changed, actually. If you look in the backend, the current definition for this just checks whether the state name currently exists: https://github.com/AllanYangZhou/oppia/blob/suggestions/core/domain/exp_services.py#L1448

seanlip replied:
(Also, may want to change the name to 'isSuggestionValid'.)

Note that, with the current implementation, only one suggestion may be accepted at a time (and the rest will all go out of date immediately.)

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,237 @@
+        backdrop: true,
+        size: 'lg',
+        resolve: {
+          isSuggestionOpen: function() {
```
Suggest: suggestionIsOpen, suggestionIsCurrent, unsavedChangesExist to indicate that these are nouns and not functions (whose names generally begin with verbs).

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -656,6 +656,16 @@ oppia.config(['$provide', function($provide) {
+      var componentName = customizationArgSpecs[0].name;
+      // The image component only works from the editor side.
+      // Prevent suggestors from trying to edit or add images.
+      if (componentName === 'filepath' && pageContext !== 'editor') {
```
Can we use the PAGE_CONTEXT.EDITOR constant instead?

seanlip replied:
Also, should we hide the image in the toolbar if the page context is wrong? It seems a bit odd to show it and then immediately show an alert if the user clicks on it.

AllanYangZhou replied:
Done. Disabling button instead of hiding the button entirely: 829581e

seanlip commented:
Hi @AllanYangZhou -- I took another pass. Thanks!

seanlip commented:
Also, /cc @jacobdavis11. Jacob, would you mind doing some functionality testing of the suggestions feature?

seanlip commented on core/templates/dev/head/player/PlayerServices.js:
```diff
@@ -489,6 +495,63 @@ oppia.controller('LearnerLocalNav', [
+          value: result.data.suggestionContent
+        }
+      }).success(function(res) {
+        $modal.open({
```
Can we do this immediately rather than waiting for the post() to return? There's an awkward pause after the submission but before the follow-up modal opens.

Please also check what would happen if there was some kind of server error -- a yellow warning message should pop up. (I.e. there should be some kind of error handler.)

AllanYangZhou replied:
Done: bf0ce27

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,237 @@
+              'for an outdated version of this exploration. It cannot be ' +
+              'accepted.';
+            var UNSAVED_CHANGES_MSG = 'You have unsaved changes to ' +
+              'this exploration. Please discard your unsaved changes if you ' +
```
Please save/discard your changes before acting on this suggestion.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/feedback_tab.html:
```diff
@@ -173,21 +175,28 @@
-    <button class="btn btn-danger" ng-click="rejectSuggestion()" ng-disabled="!(activeThread.suggestion.status == 'new')">Reject</button>
-    <button class="btn btn-success" ng-click="acceptSuggestion()" ng-disabled="!(activeThread.suggestion.status == 'new')">Accept</button>
+    <button class="btn btn-danger" ng-click="rejectSuggestion()" ng-disabled="!canReject">Reject</button>
+    <button class="btn btn-success" ng-click="acceptSuggestion()" ng-disabled="!canAccept || commitMessage.length == 0">Accept</button>
```
So, I'm finding, in testing, that I can't accept anything at all; this is always disabled. Do you find this too?

One thing that makes this irritating is that there's no error message explaining why it's disabled. I think that, as a general principle, if either button is disabled, there needs to be some message telling me why.

AllanYangZhou replied:
Sorry, I just saw this, must've missed it. Is this because the commit message is empty? I can make the commit message box red if it's blank to indicate something's wrong (also, commit message label will be changed to 'Brief Description of Changes').

seanlip replied:
Oh, it could be -- I don't even remember noticing the commit message box when I was testing the functionality, so that's probably it!

seanlip commented:
@AllanYangZhou, I did a bit of functionality testing; some comments:

- When the suggestion has been accepted/rejected, I'd hide the Accept/Reject buttons on the bottom. I would also explicitly say "This suggestion has been accepted." or "This suggestion has been rejected." -- the current either/or text is confusing.

- If I am not an exploration editor, the Accept/Reject buttons should not be shown. (Note that it is possible to view the editor page even though one is not an editor of that exploration.)

- Currently, adding any new message to a suggestion thread throws an error.

(This probably also demonstrates that e2e tests are really needed!)

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -637,10 +637,10 @@ oppia.factory('rteHelperService', [
   $provide.decorator('taOptions', [
       '$delegate', '$modal', '$timeout', 'focusService', 'taRegisterTool',
-      'rteHelperService',
+      'rteHelperService', 'explorationContextService',
```
FYI, if we do this, I think we'll need to include ExplorationContextService.js at the bottom of all pages. Can you remove it from existing pages and put it in base.html instead?

AllanYangZhou replied:
OK. Just to confirm, you referring to the `{{ include_js_file }}` stuff in the html files right?

seanlip replied:
@AllanYangZhou: Yes, that's right.

AllanYangZhou replied:
Done.

sbhowmik89 commented on core/domain/feedback_services.py:
```diff
@@ -40,7 +40,7 @@ def _get_thread_dict_from_model_instance(thread):
         'subject': thread.subject,
         'summary': thread.summary,
-        'full_thread_id': thread.id}
+        'thread_id': thread.id.split('.')[1]}
```
Use get_exp_id_from_full_thread_id() instead.

AllanYangZhou replied:
Thanks, fixed.

jacobdavis11 commented:
This works very nicely, and the implementation of the rte looks solid. Some comments:

1. The "suggest a change" tooltip is split over three lines when two would seem more natural.

2*. When attempting to reply to a suggestion, after clicking Send I get the error "Warning! Suggestion thread status cannot be changed manually."

3. If possible it would be nice to have the description of the suggestion appear inline, rather than requiring the user to click "View suggestion" since the suggestion is clearly what they came here to view. That way the acceptance and rejection buttons could also be inline.

4. If a suggestion is for an earlier version of the exploration then it cannot be accepted. I understand this as a technical limitation but it seems weird then to allow the user to review the suggestion before forcing them to reject it. I suggest either automatically rejecting such suggestions or making the messaging that they are unusable more prominent - at least have it appear before clicking "View suggestion".

5. I suggest omitting the "Accepted suggestion by:" at the start of the commit messages.

6. I suggest having the suggester write the commit message (then give the reviewer an option but not obligation to change it), and then using that as the title of the suggestion thread, rather than having them all be called "Suggestion from a learner".

7. When I accept a change it is not immediately added on the history tab, I have to refresh the page to see it. This contrasts with the behaviour when editing, in which the history to is updated immediately after each change is saved.

8. Is there a plan for allowing outdated suggestions to be accepted? Currently it's rather limiting as for example given a block of suggestions you can accept most one of them.

9. For old suggestions you say "This suggestion has already been accepted or rejected"; it would be nice to specify which of these has happened.

10. (@seanlip) I started all the jobs on the admin page but replies to the suggestor still didn't appear on their notifications page. Neither did replies to regular feedback messages. Is there something else I need to do to make this happen?

seanlip commented:
@jacobdavis11 -- thanks for the feedback!

FYI, for _3, I don't think we'll do this; suggestions can get quite bulky. For _4, I've suggested to @AllanYangZhou that we change the definition of 'can be accepted' to 'a state with this name still exists', which is what we actually do in the backend at the moment. For _6, do you imagine suggesters will know what a commit message is? I don't think they'll write one or have any sense of where it's going to be used, so I disagree with your suggestion here. I agree with all other points.

Re _10, I think all you need to do is need to start the DashboardRecentUpdatesAggregator continuous computation, and let that run for at least one cycle. I did that last night while testing something else and it seemed to work fine, though I haven't tested it with messages on suggestions since I couldn't actually add any due to the error you mentioned in _2.

jacobdavis11 commented:
_3. The current change comparison view seems to be of a fixed width so couldn't you just incorporate it into the main view? In terms of length it shouldn't in general be too bad as it's just a single state content. And it doesn't matter if the user has to scroll down to see the discussion, as that is generally less important.

_6. You don't have to use that term, you can call it something like "Brief description of changes". I think there's a lot of value in having it come from the suggester. Actually you might want to avoid using the term "commit message" anywhere, as it's a bit technical; Wikipedia uses "edit summary" instead.

_10. Thanks, I probably did something wrong.

seanlip commented:
_3. Perhaps; I'm not that keen on tackling it now, though, given that there are quite a few other things to tackle and the current functionality seems fine to me. I think we can leave it for a future iteration.

_6. Hm, fair points. @AllanYangZhou, what do you think?

AllanYangZhou commented:
_6. I think it's a great idea. I'll add a box for learners to write a brief description of their changes with their suggestion, and use that as the subject line. It will also be the default commit message (or 'Description of Changes') when the editor accepts the suggestion (the editor can choose to edit it).

jacobdavis11 commented:
_3. I would strongly suggest doing this as it makes the suggestion review process much more fluid, and it's important that we encourage rapid responsiveness to suggestions. I would have thought you could pretty much just move the comparison html from the modal into the main view. Are there difficulties that I'm missing?

seanlip commented:
My main concern is that it might be tricky/time-consuming from an implementation perspective, but I'll let @AllanYangZhou make the call since he's doing the implementation (and he may also notice UI issues I hadn't thought of).

AllanYangZhou commented:
Yeah, I think it's probably best to leave _3 and finish up and merge everything else first. Moving the current modal code into the main view might be a bit tedious.

I also suspect the current modal design might not look as good in the main view (it's not as wide, and the buttons in their current position might get pushed off the end of the screen) so I would also need to think about redesigning what it looks like.

seanlip commented:
@AllanYangZhou -- just checking, should I review, or do you have more changes coming in?

AllanYangZhou commented:
@seanlip Not quite done yet.

BTW, you mentioned in an earlier that non-editors can see the feedback page as well. Could you elaborate further? I was under the impression a person would have to be given permissions to see the feedback page and things like that.

seanlip commented:
Ah -- nope, they just can't edit. Go to /create/{{exploration_id}} when logged out.

AllanYangZhou commented:
OK I see. I'll use editabilityService to check before displaying accept/reject buttons then.

AllanYangZhou commented:
@jacobdavis11 Thanks for all the feedback, I've gone through and address most of the points:

_1. "Suggest a Change" (3 lines) is now "Suggest Changes" (2 lines)

_2. Fixed.

_3. Deferred.

_4. Outdated suggestions is no longer a problem.

_5. 'Accepted suggestion by:' prefix has been removed.

_6. I implemented a system to have suggesters write a brief description of their suggestion, which is now used as the subject of the resulting suggestion thread and as the default commit message for the reviewer (which they can edit if they want). Also, I changed "Commit Message" to "Brief description of changes".

_7. Thanks for pointing this out! This should be fixed now.

_8. See _4.

_9. Implemented.

seanlip commented:
Thanks, @AllanYangZhou! 

@jacobdavis11 -- would you mind doing another functionality test for this?

Also, @AllanYangZhou -- just let me know whenever you'd like me to review the code. Thanks!

AllanYangZhou commented:
@seanlip Yup!

One thing I know is still here is (at least in Chrome) if there is an image in the content and you try to manipulate it while suggesting (not actually editing it, since that's prohibited, but just moving it around) causes strange errors with textangular-rangy. I might need a little bit of time to figure this one out...

seanlip commented:
Does it happen in other RTEs, or just when suggesting? If the former, I would say, don't handle this now, but file it as a separate bug.

jacobdavis11 commented:
This looks great! All the issues I mentioned are fixed and it works very smoothly. A couple of extra things I noticed:

_11. Create an exploration that has hypens in its exploration ID and publish it. Go to the explore url for that exploration. You will see error "Warning! Parameter values passed into interpolateUrl must only contain alphanumerical characters or spaces: '4PBe1P1P-xrx'". This seems unrelated to what you are doing, but it doesn't occur in develop. This may be due to an error that has since been fixed in develop, so I suggest you merge develop into your branch and see if the problem still occurs.

_12. Create and publish an exploration, have a different user make a suggestion, and have the exploration author reply to the suggestion. When the suggester goes to their dashboard and clicks on the suggestions thread for the first time they are taken to the exploration editor but are captured by the tutorial, which makes it seem like they have been accidentally taken to the wrong place. Even after having dismissed the tutorial, future clicks on the message thread only go the exploration editor's feedback tab, rather than to the message itself. Would it be possible to add internal urls for the individual messages and then link from the dashboard directly through to them?

seanlip commented:
Hi @AllanYangZhou -- FYI, I just merged oppia:develop into oppia:suggestions. If you merge that into your branch, hopefully that will take care of _11. Thanks!

AllanYangZhou commented:
@jacobdavis11 
_12: Good idea! This is more of a general feedback problem than it is a suggestions one, so I'll file a separate issue.

AllanYangZhou commented:
@seanlip Fixed all the tests and addressed the remaining points, PTAL!

AllanYangZhou commented:
@seanlip I've finished the merge (whew!).

seanlip commented on core/controllers/feedback_test.py:
```diff
@@ -363,7 +360,7 @@ def setUp(self):
         # Login and create exploration and suggestions.
         self.login(self.EDITOR_EMAIL)
-        response = self.testapp.get('/create/%s' % self.EXP_ID)
+        response = self.testapp.get('/explore/%s' % self.EXP_ID)
```
It's odd to do this here; it's better to do it just above line 388, where the csrf_token is used. (Especially since the next few instructions below have nothing to do with the learner view.)

AllanYangZhou replied:
Moved.

seanlip commented on core/controllers/feedback_test.py:
```diff
@@ -462,14 +465,12 @@ def test_create_and_fetch_suggestions(self):
                           thread_id))
-        self.assertEqual(
-            response_dict['suggestion']['id'], full_thread_id)
+        self.assertEqual(feedback_services.get_thread_id_from_full_thread_id(
```
Shouldn't the response dict that's being sent to the frontend just use the thread id (and not the concatenation of the exp and thread ids)?

AllanYangZhou replied:
I changed the list response (the list of threads when first loading the feedback tab) to return only thread_id. However, the response to loading a particular thread includes the suggestion's id, which is the full_thread_id. This value isn't actually used anywhere, so I didn't feel the need to change it.

seanlip replied:
Hm. In that case, if it's not actually used anywhere, do we need to include it in the dict at all?

AllanYangZhou replied:
Yeah, I've removed it, and replaced the test to check description instead.

seanlip commented on core/controllers/feedback_test.py:
```diff
@@ -515,11 +516,9 @@ def _accept_suggestion(self, thread_id, csrf_token,
     def _reject_suggestion(self, thread_id, csrf_token,
```
If I'm not mistaken, what's being done below is to generate the full thread id from the exp id and thread id, then take just the thread id bit here. Wouldn't it be better to just use the thread id directly?

Ditto for accept_suggestion, above. In general I would suggest minimizing the use of full_thread_id where possible -- it becomes particularly confusing when it's called thread id.

AllanYangZhou replied:
Fixed.

seanlip commented on core/domain/exp_services.py:
```diff
@@ -1461,17 +1461,18 @@ def accept_suggestion(editor_id, thread_id, exploration_id, commit_message):
+            suggestion_author_username)
         update_exploration(
-            editor_id, exploration_id, change_list, full_commit_message)
+            suggestion_author_id, exploration_id, change_list,
```
Hang on -- shouldn't the author id here be the editor_id (since they're actually making the commit), with the author username prepended to the commit message?

I wonder if, in cases like this, we should also store the suggestion_author_id in a separate field in the commit history. What do you think?

AllanYangZhou replied:
I've fixed it to use the editor_id, and put back the 'Accepted suggestion by: '

seanlip commented on core/domain/exp_services.py:
```diff
@@ -1461,17 +1461,18 @@ def accept_suggestion(editor_id, thread_id, exploration_id, commit_message):
-        thread = feedback_models.FeedbackThreadModel.get_by_exp_and_thread_id(
-            exploration_id, thread_id)
-        thread.status = feedback_models.STATUS_CHOICES_FIXED
+        thread = (feedback_models.FeedbackThreadModel.
```
Doesn't look like you need this and the next two lines at all ... you just get the model and then put it, without changing anything.

AllanYangZhou replied:
Removed.

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -50,6 +45,9 @@ def test_feedback_ids(self):
         datastore_id = feedback_models.FeedbackMessageModel.get_messages(
             exp_id, thread_id)[0].id
+
+        full_thread_id = (feedback_models.FeedbackThreadModel
```
I don't understand -- do you use this anywhere?

seanlip replied:
Sorry! My mistake -- I didn't see there was a cutoff in the github view of the PR below this one. Please ignore this comment.

seanlip commented on core/domain/feedback_services_test.py:
```diff
@@ -180,9 +179,8 @@ def test_get_open_threads_with_suggestions(self):
-            threads[0]['full_thread_id'],
-            feedback_models.FeedbackThreadModel.generate_full_thread_id(
-                self.EXP_ID1, self.THREAD_ID1))
+            threads[0]['thread_id'],
```
Optional: all this (and the next line) can go on the line above.

AllanYangZhou replied:
Fixed.

seanlip commented on core/domain/user_jobs_continuous.py:
```diff
@@ -204,8 +204,11 @@ def map(item):
             for thread in threads:
-                if thread['full_thread_id'] not in feedback_thread_ids_list:
-                    feedback_thread_ids_list.append(thread['full_thread_id'])
+                full_thread_id = (feedback_models.FeedbackThreadModel
```
FYI, we generally do:

    full_thread_id = (
        feedback_models.FeedbackThreadModel.generate_full_thread_id(
            exp_model.id, thread['thread_id']))

This gives more space on the right.

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -260,11 +260,13 @@ class SuggestionModel(base_models.BaseModel):
     exploration_version = ndb.IntegerProperty(required=True, indexed=True)
     # Name of the corresponding state.
     state_name = ndb.StringProperty(required=True, indexed=True)
+    # Learner provided description of suggestion changes.
```
Learner-provided

AllanYangZhou replied:
Done.

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -260,11 +260,13 @@ class SuggestionModel(base_models.BaseModel):
     state_name = ndb.StringProperty(required=True, indexed=True)
+    # Learner provided description of suggestion changes.
+    description = ndb.StringProperty(required=True, indexed=True)
     state_content = ndb.JsonProperty(required=True, indexed=False)
```
Can you add a comment describing this field (and its format), too?

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/base.html:
```diff
@@ -278,8 +278,9 @@
 
-      {{ include_js_file('components/ExplorationSummaryTileDirective.js') }}
       {{ include_js_file('components/ObjectEditorDirective.js') }}
+      {{ include_js_file('services/explorationContextService.js') }}
```
Consider following the existing pattern, and grouping services/... with the block above this one.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,256 @@
-          $modalInstance.close({
-            action: 'reject'
+    var _isSuggestionValid = function() {
+      return explorationStatesService.getStates()[
```
It would be cleaner to add a hasState() method to explorationStatesService (in editor/EditorServices.js) and call that instead, I think. The hasState() method does:

    return _states.hasOwnProperty(stateName);

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,256 @@
+      return (changeListService.getChangeList().length > 0);
+    };
+
+    $scope.viewSuggestionBtnType = function() {
```
s/Btn/Button

seanlip replied:
Also -- how about getSuggestionButtonType()? Clearer than 'view', I think, from the point of view of the caller.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,256 @@
+          description: function() {
+            return $scope.activeThread.suggestion.description;
+          },
+          oldContent: function() {
```
Interesting -- thinking about it, this is more like currentContent, right? (E.g. if the modal is opened after the suggestion has been accepted.)

AllanYangZhou replied:
Makes sense, changed to currentContent.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,256 @@
+            var SUGGESTION_REJECTED_MSG = 'This suggestion has already been ' +
+              'rejected.';
+            var SUGGESTION_INVALID_MSG = 'This suggestion was made ' +
+              'for a state that no longer exists. It cannot be  accepted.';
```
remove one of the spaces between 'be' and 'accepted'

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,256 @@
+            }
+
+            $scope.oldContent = (
+              oldContent || '<span style="color:red">Oops! This state no ' +
```
Can you do this in the HTML template? It's a bit ugly (and can be a security risk) to manipulate HTML in the JS.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/FeedbackTab.js:
```diff
@@ -19,162 +19,256 @@
+            $scope.acceptSuggestion = function() {
+              $modalInstance.close({
+                action: ACTION_ACCEPT_SUGGESTION,
+                commitMsg: $scope.commitMessage
```
Mildly prefer commitMessage

AllanYangZhou replied:
Changed.

seanlip commented on core/templates/dev/head/editor/feedback_tab/ThreadDataService.js:
```diff
@@ -50,11 +55,12 @@ oppia.factory('threadDataService', [
-        if (_data.threadList[i].thread_id === threadId) {
-          _data.threadList[i].messages = data.messages;
+    $http.get(_THREAD_HANDLER_PREFIX + threadId).success(function(data) {
+      var combined = _data.feedbackThreads.concat(_data.suggestionThreads);
```
Hmm. s/combined/allThreads, maybe?

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/ThreadDataService.js:
```diff
@@ -81,14 +87,15 @@ oppia.factory('threadDataService', [
     },
-    addNewMessage: function(threadId, newMessage, newStatus, successCallback, errorCallback) {
+    addNewMessage: function(
+      threadId, newMessage, newStatus, successCallback, errorCallback) {
```
Optional: we tend to indent lines like this an additional 2 spaces to distinguish from the next line. Ditto for line 129.

(It's OK either way since the linter doesn't catch it, but I thought I'd mention it since you might find it easier to read, too.)

seanlip commented on core/templates/dev/head/editor/feedback_tab/ThreadDataServiceSpec.js:
```diff
@@ -91,10 +90,12 @@ describe('retrieving threads service', function() {
     for (var i = 0; i < mockFeedbackThreads.length; i++) {
-      expect(threadDataService.data.threadList).toContain(mockFeedbackThreads[i]);
+      expect(threadDataService.data.feedbackThreads).toContain(
+          mockFeedbackThreads[i]);
```
should be only +2 spaces indentation, ditto below

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/feedback_tab.html:
```diff
@@ -93,9 +94,11 @@
-                  <button class="btn btn-primary" ng-if="activeThread.suggestion && $index==0" ng-click="showSuggestionModal()">
-                    View Suggestion
-                  </button>
+                  <div ng-if="activeThread.suggestion && $index==0">
```
add spaces around ==

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/feedback_tab.html:
```diff
@@ -173,21 +176,28 @@
   </div>
 
   <div class="modal-footer">
+    <div ng-show="errorMessage" style="text-align: right; width: 100%; color: red; margin-bottom: 10px;">
```
Styling here is complicated enough that I'd suggest making it into a class.

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/editor/feedback_tab/thread_table_directive.html:
```diff
@@ -11,7 +11,7 @@
       </td>
       <td class="col-lg-6 col-md-6 col-sm-6">
-        <div><[thread.subject]></div>
+        <div><[thread.subject | limitTo: 35]><[thread.subject.length > 35 ? '&hellip;' : '']></div>
```
We do have a 'truncate' filter, FYI...

AllanYangZhou replied:
Changed to use filter.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -756,6 +757,10 @@ oppia.config(['$provide', function($provide) {
           tooltiptext: componentDefn.tooltip,
+          disabled: function() {
+            // Check canUseFs, a property on the textAngular directive scope.
+            return !this.$parent.$parent.canUseFs && componentDefn.requiresFs;
```
Eww (to $parent.$parent). Do you think this should be a globally encapsulated in a service instead (maybe part of the explorationContextService)?

AllanYangZhou replied:
Fixed turns out I didn't need $parent.$parent anyways.

seanlip commented on core/templates/dev/head/forms/formBuilder.js:
```diff
@@ -847,66 +861,77 @@ oppia.config(['$provide', function($provide) {
+          // Currently, operations affecting the filesystem are allowed only in
+          // the editor context.
+          $scope.canUseFs = (
+            explorationContextService.getPageContext() === PAGE_CONTEXT.EDITOR);
```
Ah -- so this is where it's defined. Is it possible to actually refer to the service/constant from within the provider, too?

seanlip replied:
And then, maybe you won't need it here at all...

AllanYangZhou replied:
I thought I had a good reason for using $parent.$parent instead, but just using explorationContextService from the provider seems to work (and I can't remember what the reason was, I was probably just confused).

AllanYangZhou replied:
I've removed $parent.$parent and am just using explorationContextService instead.

seanlip commented on core/templates/dev/head/player/LearnerLocalNav.js:
```diff
@@ -19,7 +19,73 @@
+
+            $scope.submitSuggestion = function() {
+              $modalInstance.close({
+                data: {
```
You don't really need the outer 'data' dict, right?

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/domain/exploration/ExplorationObjectFactory.js:
```diff
@@ -117,6 +117,10 @@ oppia.factory('ExplorationObjectFactory', [
       return this.getState(this.initStateName);
     };
 
+    Exploration.prototype.getContentHtml = function(stateName) {
```
I just realized that we should probably call this something like Exploration.prototype.getUninterpolatedContentHtml, since it may contain parameters. (This is as opposed to the HTML that is eventually displayed as the content.)

AllanYangZhou replied:
Done.

seanlip commented on core/templates/dev/head/player/exploration_player.html:
```diff
@@ -33,6 +33,55 @@
 {% endblock navbar_breadcrumb %}
 
 {% block local_top_nav_options %}
+  <div ng-controller="LearnerLocalNav">
```
Why two of these (here and in the line below)? Also note the inclusion of learner_local_nav.html on line 85; this should be moved into there.

AllanYangZhou replied:
Oops, must be an artifact of the merge. I've fixed it.

seanlip commented on core/tests/gae_suite.py:
```diff
@@ -22,10 +22,10 @@
 import os
 import sys
 import unittest
+import argparse
```
The location of this and other similar imports shouldn't change.

seanlip replied:
Actually, it's not worth spending a lot of time on this one, I think. For the 'argparse' imports (and only those), I would add a ```# pylint: disable``` pragma (see elsewhere in the code for how this is done).

AllanYangZhou replied:
So the error pylint is giving me is that the other imports should come before argparse. That is, it is all the other imports that are flagged by pylint, not the `import argparse` line itself. So I would have to add a `pylint disable` to every non-argparse import.

AllanYangZhou replied:
Done: a710465

seanlip commented on core/storage/feedback/gae_models.py:
```diff
@@ -260,11 +260,15 @@ class SuggestionModel(base_models.BaseModel):
     # Name of the corresponding state.
     state_name = ndb.StringProperty(required=True, indexed=True)
+    # Learner-provided description of suggestion changes.
+    description = ndb.StringProperty(required=True, indexed=True)
```
Can you change this to TextProperty and make it "indexed=False"? StringProperty takes max 500 chars.

seanlip replied:
Sorry, I missed this -- can you change this to TextProperty and make it "indexed=False"? StringProperty takes max 500 chars.

AllanYangZhou replied:
Done.

seanlip commented:
LGTM. Thanks @AllanYangZhou!
